File: avoidance/biopatrec_comm_test.m
function test_tl()

% This is used to test the communication to ArduPilot in simulation
% DO NOT USE THIS IN A REAL VEHICLE

t = tcpip('127.0.0.1', 8080, 'NetworkRole', 'Client');
fopen(t)
flushinput(t)

%
% Motors:
%  Move North/South     1,A
%  Turn Right/Left      2,B
%  Move Up/Down         3,C
%  Move East/West       4,D
%

REST    = 0;
PITCH   = 'A';
YAW     = 'B';
THRUST  = 'C';
ROLL    = 'D';

HALT    = 'H';
ARM     = 'T';

% Test the connection
fwrite(t, 'AC');
rsp = char(fread(t, 1));
if(rsp == 'C')
    fprintf('Connection Successful\n');
else
    fprintf('Conn Test Failure\n')
    fprintf('RSP: %d,0x%X\n', rsp, rsp)
    fclose(t)
    return
end

% Wait up to 45 seconds for initialization 
for i = 1:45
    fwrite(t, ['M',0,REST,0,0]);
    rsp = fread(t, 1)
    if(rsp == ARM)
        pause(1)
    elseif(rsp == REST)
        break
    else
        fprintf('Initialization Failure: %d\n', i)
        fprintf('RSP: %d,0x%X\n', rsp, rsp)
        fclose(t)
        return
    end
end
fprintf('Initialization Successful\n');

% Move North
fprintf('Moving North\n');
for i = 1:5
    fwrite(t, ['M',0,PITCH,1,0])
    rsp = fread(t, 1);
    if(isempty(rsp) | rsp ~= PITCH)
        fprintf('Invalid response %d\n', i)
        fprintf('RSP: %d,0x%X\n', rsp, rsp)
        fclose(t)
        return
    end
    pause(1)
end

% Turn east
fprintf('Turning East\n');
for i = 1:5
    fwrite(t, ['M',0,YAW,1,0])
    rsp = fread(t, 1);
    if(isempty(rsp) | rsp ~= YAW)
        fprintf('Invalid response %d\n', i)
        fprintf('RSP: %d,0x%X\n', rsp, rsp)
        fclose(t)
        return
    end
    pause(1)
end

% Move east
fprintf('Moving East\n');
for i = 1:5
    fwrite(t, ['M',0,ROLL,1,0])
    rsp = fread(t, 1);
    if(isempty(rsp) | rsp ~= ROLL)
        fprintf('Invalid response %d\n', i)
        fprintf('RSP: %d,0x%X\n', rsp, rsp)
        fclose(t)
        return
    end
    pause(1)
end

% Move west
fprintf('Moving West\n');
for i = 1:5
    fwrite(t, ['M',0,ROLL,0,0])
    rsp = fread(t, 1);
    if(isempty(rsp) | rsp ~= ROLL)
        fprintf('Invalid response %d\n', i)
        fprintf('RSP: %d,0x%X\n', rsp, rsp)
        fclose(t)
        return
    end
    pause(1)
end

fprintf('Landing\n');
for i = 1:45
    fwrite(t, ['M',0,THRUST,0,0])
    rsp = fread(t, 1);
    if(rsp == 'H')
        fprintf('Vehicle Landed\n')
        break
    elseif(rsp == THRUST)
        pause(1)
    else
        fprintf('Landing failure %d\n')
        fprintf('RSP: %d,0x%X\n', rsp, rsp)
        fclose(t)
        return
    end
end

fclose(t)
fprintf('Success!\n')

end

File: avoidance/biopatrec_comm_test.py
#!/usr/bin/env python

# This is used to test the communication to ArduPilot in simulation
# DO NOT USE THIS IN A REAL VEHICLE

import socket, sys
import time, signal

TCP_IP = '127.0.0.1'
TCP_PORT = 8080
BUFFER_SIZE = 1024
FPS = 5
landed = False

REST        = '\0'
PITCH       = 'A'
YAW         = 'B'
ALT         = 'C'
ROLL        = 'D'
WAIT        = 'T'
HALT        = 'H'

REST_CMD    = bytearray(['M',0,REST,1,0])
NORTH_CMD   = bytearray(['M',0,PITCH,1,0])
SOUTH_CMD   = bytearray(['M',0,PITCH,0,0])
EAST_CMD    = bytearray(['M',0,ROLL,1,0])
WEST_CMD    = bytearray(['M',0,ROLL,0,0])
RIGHT_CMD   = bytearray(['M',0,YAW,1,0])
LEFT_CMD    = bytearray(['M',0,YAW,0,0])
UP_CMD      = bytearray(['M',0,ALT,1,0])
DOWN_CMD    = bytearray(['M',0,ALT,0,0])

def signal_handler(signal, frame):
    print("Caught signal: Closing")
    s.close()
    sys.exit(0)
signal.signal(signal.SIGINT, signal_handler)

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((TCP_IP, TCP_PORT))

print("Testing Connection...")
s.send('A')
time.sleep(0.1)
s.send('C')
rsp = s.recv(1)
if rsp != 'C':
    print("Conn test failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
    s.close()
    sys.exit(1)
print("Connected!")

print("Sending Rest")
for i in range(45):
    s.sendall(REST_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp == WAIT:
        time.sleep(1)
    elif rsp == REST:
        print("System ready")
        break
    else:
        print("Init failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)

print("Moving North")
for i in range(20):
    s.sendall(NORTH_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != PITCH:
        print("North failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Moving South")
for i in range(20):
    s.sendall(SOUTH_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != PITCH:
        print("South failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Moving East")
for i in range(20):
    s.sendall(EAST_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != ROLL:
        print("East failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Moving West")
for i in range(20):
    s.sendall(WEST_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != ROLL:
        print("West failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Turning Right")
for i in range(20):
    s.sendall(RIGHT_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != YAW:
        print("Right failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Turning Left")
for i in range(20):
    s.sendall(LEFT_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != YAW:
        print("Left failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Moving Up")
for i in range(5):
    s.sendall(UP_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp != ALT:
        print("Up failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

print("Moving Down")
for i in range(60):
    s.sendall(DOWN_CMD)
    rsp = s.recv(BUFFER_SIZE)
    if rsp == HALT:
        print("System landed")
        landed = True
        break
    elif rsp != ALT:
        print("Down failure: Received: %d/0x%X" % (ord(rsp),ord(rsp)))
        s.close()
        sys.exit(1)
    time.sleep(1.0/FPS)

if landed:
    print("Success")
else:
    print("Not landed!")
s.close()
sys.exit(0)
